File: 2022.jl/03.jl
"""
Day 3: Rucksack Reorganization
https://adventofcode.com/2022/day/3
"""

function part1(lines)
    priorities = Vector{Int}()

    for line in lines
        first_compartment = line[1:length(line)÷2]
        second_compartment = line[length(line)÷2+1:end]
        letter = intersect(first_compartment, second_compartment)[1]
        priority = Int(letter)
        if priority > 96
            priority -= 96
        else
            priority -= 38
        end
        push!(priorities, priority)
    end

    return sum(priorities)
end

function part2(lines)
    priorities = Vector{Int}()

    for i in 1:3:length(lines)
        letter = intersect(lines[i], intersect(lines[i+1], lines[i+2]))[1]
        priority = Int(letter)
        if priority > 96
            priority -= 96
        else
            priority -= 38
        end
        push!(priorities, priority)
    end

    return sum(priorities)
end

example = readlines("examples/03.txt")
input = readlines("inputs/03.txt")

@assert part1(example) == 157
println(part1(input))

@assert part2(example) == 70
println(part2(input))

File: 2025.zig/03.zig
// Day 3: Lobby
// https://adventofcode.com/2025/day/3

const std = @import("std");

fn part1(path: []const u8) !u64 {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    var file_buffer: [4096]u8 = undefined;
    var reader = file.reader(&file_buffer);
    var sum: u64 = 0;
    while (try reader.interface.takeDelimiter('\n')) |line| {
        var a: u8 = 0;
        var b: u8 = 0;
        var index: usize = 0;
        for (line, 0..) |c, i| {
            if (c - 48 > a) {
                a = c - 48;
                index = i;
            }
        }
        if (index == line.len - 1) {
            for (line[0..(line.len - 1)]) |c| {
                if (c - 48 > b) {
                    b = c - 48;
                }
            }
            sum += b * 10 + a;
        } else {
            for (line[index + 1 ..]) |c| {
                if (c - 48 > b) {
                    b = c - 48;
                }
            }
            sum += a * 10 + b;
        }
    }
    return sum;
}

fn part2(path: []const u8) !u64 {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    var file_buffer: [4096]u8 = undefined;
    var reader = file.reader(&file_buffer);
    var sum: u64 = 0;
    while (try reader.interface.takeDelimiter('\n')) |line| {
        var joltage = std.mem.zeroes([12]u8);
        for (line, 0..) |c, i| {
            const len: i32 = @intCast(line.len);
            const pos: i32 = @intCast(i);
            for (joltage[@max(0, 12 - (len - pos))..], @max(0, 12 - (len - pos))..) |n, j| {
                if (c - 48 > n) {
                    joltage[j] = c - 48;
                    for ((j + 1)..12) |x| {
                        joltage[x] = 0;
                    }
                    break;
                }
            }
        }
        var number: usize = 0;
        for (joltage, 0..) |n, i| {
            number += std.math.pow(usize, 10, (11 - i)) * n;
        }
        sum += number;
    }
    return sum;
}

pub fn main() !void {
    std.debug.assert(try part1("./examples/03.txt") == 357);
    std.debug.print("{d}\n", .{try part1("./inputs/03.txt")});
    std.debug.assert(try part2("./examples/03.txt") == 3121910778619);
    std.debug.print("{d}\n", .{try part2("./inputs/03.txt")});
}
